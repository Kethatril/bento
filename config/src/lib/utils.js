import objectPath from "object-path";
import { parse as shoelace } from "@kenvunz/shoelace";
import { join, resolve as res, dirname } from "path";
import bl from "browserslist";
import { existsSync, readFileSync } from "fs";
import querystring from "querystring";
import merge from "lodash/merge";

export function enhance(options) {
    return objectPath(options);
}

export function resolve(options) {
    ["input", "output"].forEach(key => {
        if (
            !options.paths ||
            !options.paths[key] ||
            options.paths[key].indexOf(options.paths.root) === 0
        )
            return;

        options.paths[key] = res(join(options.paths.root, options.paths[key]));
    });

    return options;
}

export function browsersync(options) {
    if (!options.paths || !options.paths.input) return options;
    options.browsersync = merge(
        browsersyncrc(options.paths.input),
        options.browsersync || {}
    );
    return options;
}

export function browserslist(options) {
    if (!options.paths || !options.paths.input) return options;
    options.browserslist = merge(
        browserslistrc(options.paths.input),
        options.browserslist || {}
    );
    return options;
}

export function babel(options) {
    if (!options.paths || !options.paths.input) return options;
    options.babel = merge(babelrc(options.paths.input), options.babel || {});
    return options;
}

export function transform(
    options,
    fns = [resolve, browserslist, browsersync, babel, shoelace, enhance]
) {
    for (const fn of fns) {
        options = fn(options);
    }

    return options;
}

export function select(options) {
    return function(cases) {
        return cases[options.get("env.value")];
    };
}

export function caught(e) {
    if (process.env.NODE_ENV === "test") throw e;
    console.error(e);
}

function browsersyncrc(path) {
    return rc(path, [
        "bs-config.js",
        ".browsersyncrc",
        "package.json?key=browsersync"
    ]);
}

function babelrc(path) {
    return rc(path, [".babelrc", "package.json?key=babel"]);
}

function browserslistrc(path) {
    const config = bl.findConfig(path);
    if (!config) return bl.defaults;

    if (config[process.env.NODE_ENV]) return config[process.env.NODE_ENV];
    return config.defaults;
}

function rc(path, files) {
    for (const item of files) {
        const [name, query] = item.split("?");
        const file = join(path, name);
        if (!existsSync(file)) continue;
        let options;
        if (query) options = querystring.parse(query);

        const object = JSON.parse(readFileSync(file));

        if (name !== "package.json") return object;
        return options && options.key ? object[options.key] : options;
    }

    const up = dirname(path);

    if (up !== path) {
        return rc(up, files);
    }
}
