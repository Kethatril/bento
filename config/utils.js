const { Config, environment } = require("webpack-config"),
    path = require("path"),
    objectPath = require("object-path");

let model;

module.exports.config = function(fn) {
    const config = new Config();
    return fn(config, environment);
};

module.exports.select = function(cases) {
    return cases[model.get("env.value")];
};

module.exports.read = function(key, defaultValue) {
    return model.get(key, defaultValue);
};

module.exports.set = function(opts) {
    let json = JSON.stringify(opts);
    const pre = objectPath(opts),
        matches = json.match(/"@{([\w\.]+)}["\/]/g);

    // replace any "@key" occurrences with the real value
    matches.forEach(match => {
        const key = match.replace(/[@{}\/"]/g, ""),
            value = pre.get(key);

        json = json.replace(
            new RegExp(match, "g"),
            match.substr(-1) === "/" ? `"${value}/` : JSON.stringify(value)
        );
    });

    model = objectPath(JSON.parse(json));
};

module.exports.format = function(key, value) {
    if (typeof value === "string") {
        return value;
    }
    const manifest = value;
    /**
     * Hack to prepend scripts/ or styles/ to manifest keys
     *
     * This might need to be reworked at some point.
     *
     * Before:
     *   {
     *     "main.js": "scripts/main_abcdef.js"
     *     "main.css": "styles/main_abcdef.css"
     *   }
     * After:
     *   {
     *     "scripts/main.js": "scripts/main_abcdef.js"
     *     "styles/main.css": "styles/main_abcdef.css"
     *   }
     */
    Object.keys(manifest).forEach(src => {
        const sourcePath = path.basename(path.dirname(src));
        const targetPath = path.basename(path.dirname(manifest[src]));
        if (sourcePath === targetPath) {
            return;
        }
        manifest[`${targetPath}/${src}`] = manifest[src];
        delete manifest[src];
    });
    return manifest;
};
