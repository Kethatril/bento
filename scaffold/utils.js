const pathIsInside = require("path-is-inside"),
    path = require("path");

module.exports.select = function(key, presets) {
    return function(answers) {
        return presets[answers.kind][key];
    };
};

module.exports.inside = function(name, root) {
    return function(value) {
        const resolved = path.resolve(value);
        const result = pathIsInside(resolved, root);
        return (
            result ||
            `${name} must be inside ${chalk.bold(
                root
            )}, it's currently resolved to ${chalk.bold(resolved)}`
        );
    };
};

module.exports.trailing = function(char) {
    return function(value) {
        if (value.substr(-1) !== char) value += "/";

        return value;
    };
};
