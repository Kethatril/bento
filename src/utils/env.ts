export interface EnvSelectOptions {
    [env: string]: any;
}

export function selector(env?: string): () => any {
    return function(choice: EnvSelectOptions = {}) {
        return choice[env]
            ? choice[env]
            : choice.default ? choice.default : null;
    };
}
