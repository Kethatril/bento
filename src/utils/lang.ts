export function instantiate(Cls, args?: any[]): any {
    args.unshift(Cls);
    const C = Cls.bind.apply(Cls, args);
    return new C();
}

interface SelectorOptions {
    [key: string]: any;
}

export function selector(env?: string): (choice: SelectorOptions) => any {
    return function(choice: SelectorOptions = {}) {
        return choice[env]
            ? choice[env]
            : choice.default ? choice.default : null;
    };
}
