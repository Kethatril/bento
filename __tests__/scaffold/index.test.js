import { scaffold, tree } from "../utils";

describe("scaffold/index.js", () => {
    afterEach(() => {
        scaffold.restore();
    });

    describe("Single Page Application", () => {
        it("folder structure matches the snapshot", () => {
            return scaffold("--default spa --no-confirm").then(dir => {
                const result = tree(
                    `./ --ignore "node_modules/, .gitkeep" -a -f -l 4`,
                    dir
                );
                expect(result.stdout).toMatchSnapshot();
            });
        });
    });

    describe("Server Side Application", () => {
        it("folder structure matches the snapshot", () => {
            return scaffold("--default ssa --no-confirm").then(dir => {
                const result = tree("./ --ignore node_modules/ -f -l 4", dir);
                expect(result.stdout).toMatchSnapshot();
            });
        });
    });
});
