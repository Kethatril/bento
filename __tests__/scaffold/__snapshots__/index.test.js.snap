// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scaffold/index.js Server Side Application folder structure matches the snapshot 1`] = `
"
./
â”œâ”€â”€ /assets
|  â”œâ”€â”€ /config
|  |  â””â”€â”€ /webpack
|  |     â”œâ”€â”€ base.js
|  |     â”œâ”€â”€ blocks.js
|  |     â”œâ”€â”€ html.js
|  |     â””â”€â”€ index.js
|  â”œâ”€â”€ /fonts
|  â”œâ”€â”€ /images
|  â”œâ”€â”€ /media
|  â”œâ”€â”€ /scripts
|  |  â””â”€â”€ main.js
|  â””â”€â”€ /styles
|     â””â”€â”€ main.scss
â”œâ”€â”€ /public
|  â””â”€â”€ /assets
â””â”€â”€ webpack.config.js

directory: 10 file: 7 symboliclink: 1 

[2K[1G"
`;

exports[`scaffold/index.js Single Page Application folder structure matches the snapshot 1`] = `
"
./
â”œâ”€â”€ .babelrc
â”œâ”€â”€ .browserslistrc
â”œâ”€â”€ .browsersyncrc
â”œâ”€â”€ /app
|  â”œâ”€â”€ /config
|  |  â””â”€â”€ /webpack
|  |     â”œâ”€â”€ base.js
|  |     â”œâ”€â”€ blocks.js
|  |     â”œâ”€â”€ html.js
|  |     â””â”€â”€ index.js
|  â”œâ”€â”€ /fonts
|  â”œâ”€â”€ /images
|  â”œâ”€â”€ index.ejs
|  â”œâ”€â”€ /media
|  â”œâ”€â”€ /scripts
|  |  â””â”€â”€ main.js
|  â””â”€â”€ /styles
|     â””â”€â”€ main.scss
â”œâ”€â”€ /public
â””â”€â”€ webpack.config.js

directory: 9 file: 15 symboliclink: 1 

[2K[1G"
`;
